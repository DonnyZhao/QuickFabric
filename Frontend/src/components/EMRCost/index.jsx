import React, { Component } from 'react'
import { Grid, GridColumn as Column } from '@progress/kendo-react-grid'
import { Spinner } from '@blueprintjs/core'
import emrHealthCell from '../EMRHealth/emrHealthCell'
import './emrCost.scss'
import '../EMRHealth/emrHealth1.scss'

import { withState } from '../../utils/components/with-state.jsx';
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip } from 'recharts';

const StatefulGrid = withState(Grid);

/**
 * EMR Cost component: Table containing cost data with a final column of cost trend chart visualization.
 */
export default class EMRCost extends Component{
    account_range = '';
    constructor(props){
        super(props)
        
        this.state = {
            data: this.props.data,
            show: false,
            primaryTab: 'all'
        }

        this.renderTableData = this.renderTableData.bind(this)
        this.getTableData = this.getTableData.bind(this);
    }

    render = () => {
        return (
        <div className="emr-cost-component">
            <h2>EMR <i className="material-icon title-arrow">double_arrow</i> <span style={{ color: "#0097E4" }}>Cost</span></h2> 
            {
              this.props.fetching || this.props.data.length === 0? 
                <Spinner /> :
                this.renderTableData(this.props.data)
            }
        </div>
        )
    }

    renderTableData = data => {
      // Modifying data: Adding costs per month array as columns for table to consume.
      let modifiedData = this.getTableData(data);
      return (
        <StatefulGrid
          data={modifiedData.newData}
          style={{ "marginTop": 30 , height: "90vh", fontSize: '13px' }}
        >
          <Column field="emrGroup" title="EMR Group" width="150px" />
          <Column field="account" title="Account" width="120px" />
          <Column field="segment" title="Segment" width="100px" />
          <Column field="businessOwner" title="Business Owner" width="150px" />
          {modifiedData.costTitles.map((title, index) => {
            return (
            <Column field={title} title={title} width="80px" cell={(props) =>
              props.dataItem[title] == "0" ?
              <td>
                <span>N/A</span>
              </td> :
              <td>
                <span>${props.dataItem[title]}</span>
              </td>
            } />
            )
          })}
          <Column title="Cost Trend" width="auto" 
            cell={(props) => 
              <td>
                <LineChart width={480} height={120} data={props.dataItem.costPerMonth.reverse()}
                    >
                    <XAxis dataKey="billMonth" />
                    <YAxis />
                    <CartesianGrid strokeDasharray="3 3"/>
                    <Tooltip />
                    <Line type="monotone" dataKey="cost" stroke="#0097E4" activeDot={{r: 8}}/>
                </LineChart>
              </td>
            } 
          />
        </StatefulGrid>
      )
    }

  getTableData = data => {
    data = this.shortlistData(data.emrGroupCost);
    let newData = []
    let costTitles = []
    data.map((row, key) => {
      let vals = {
        emrGroup: row.emrGroup,
        account: row.account,
        segment: row.segment,
        businessOwner: row.businessOwner,
        costPerMonth: row.costPerMonth
      }
      row.costPerMonth.map((c) => {
        vals[c.billMonth] = c.cost;
        if (!costTitles.includes(c.billMonth))
          costTitles.push(c.billMonth)
      })
      newData.push(vals)
    })
    return {
      newData: newData,
      costTitles: costTitles
    }
  }
  
  shortlistData = data => {
    if (this.props.superAdmin) return data;
    let emrRoles = this.props.dccRoles.filter(x => x.serviceType === 'EMR')[0];

    if (emrRoles === undefined) return [];
  
    let roleSegments = [];
    emrRoles.roles.forEach((value) => {
        value.segments.forEach(x => {
          roleSegments.push(x);
        })
    });
    let roleMap = {};
    roleSegments.forEach((value, index) => {
      let accounts = [];
      value.accounts.forEach((act, index) => {
        accounts.push(act.accountId);
      })
      roleMap[value['segmentName'].toLowerCase()] = accounts;
    })
    
    let newData = data;
    newData = newData.filter((value) => {
      let type = value.segment;
      if (type in roleMap) {
        if (roleMap[type].includes(value.account)) {
          return value;
        }
      }
    });
  
    return newData;
  }
}