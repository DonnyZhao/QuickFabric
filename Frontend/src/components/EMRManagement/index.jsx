import React, { Component } from 'react'
import { Grid, GridColumn as Column, GridDetailRow, GridToolbar, GridCell } from '@progress/kendo-react-grid'
import { process } from '@progress/kendo-data-query'
import { ExcelExport } from '@progress/kendo-react-excel-export'
import { Tab, Tabs, Menu, MenuItem, Popover, Position} from '@blueprintjs/core'
import { debounce } from 'lodash';
import emrStatusCell from './emrStatusCell'
import emrStepStatusCell from './emrStepStatusCell'
import amirotationStatus from './AMIRotationDays'
import './emrManagement.css'
import SubmitActionButtonCell from './SubmitActionButtonCell'
import CreateCluster from './CreateCluster'
import { withState } from '../../utils/components/with-state.jsx';

import '@progress/kendo-theme-bootstrap/scss/all.scss';

const CLUSTER_REFRESH_TIME = 1000 * 60 * 5;

const StatefulGrid = withState(Grid);

/**
 * Component to show extra information about the cluster and creation process in the expanded version of the kendo table.
 */
export class StepsStatus extends GridDetailRow {
  constructor(props){
    super(props)
  }

  render() {
    const stepsStatusData = this.props.stepsStatusData[this.props.dataItem.clusterId]
    if(stepsStatusData && stepsStatusData.length) {
      return (
        <div className='step-status-component'>
          <Grid
            data={stepsStatusData}
            resizable={true}
            reorderable={true}
            style={{ fontSize: '13px' }}
          >
            <Column field="stepId" title="Step ID" width="auto" />
            <Column field="name" title="Step Name" width="auto" />
            <Column field="status" title="Step Status" width="auto" cell={emrStepStatusCell} />
            <Column field="stepType" title="Step Type" width="auto" />
            <Column field="creationTimestamp" title="Creation Timestamp" width="auto" />
          </Grid>
        </div>
      )
    } else {
      return (
        <div>
          <Grid
            resizable={true}
            reorderable={true}
            style={{ fontSize: '13px' }}
          >
            <Column field="stepId" title="Step ID" width="auto" />
            <Column field="name" title="Step Name" width="auto" />
            <Column field="status" title="Step Status" width="auto" cell={emrStepStatusCell} />
            <Column field="stepType" title="Step Type" width="auto" />
            <Column field="creationTimestamp" title="Creation Timestamp" width="auto" />
          </Grid></div>
      )
    }
  }
}

/**
 * EMR Management Component
 */
export default class EMRManagement extends Component {
  showCreateTab = false;
  showTerminateTab = false;
  showAddTab = false;
  showRotateTab = false;
  showTestSuitesTab = false;
  account_range = '';
  constructor(props) {
    super(props)
    this.state = {
      data: this.props.data,
      show: false,
      primaryTab: 'all',
      primaryAction: 'allClusters',
      tabHeading: 'View All Clusters',
      uiListData: this.props.uiListData,
      clusterId: ''
    }

    this.createGroupState = this.createGroupState.bind(this)
    this.dataStateChange = this.dataStateChange.bind(this)
    this.expandChange = this.expandChange.bind(this)
    this.export = this.export.bind(this)
    this.handlePrimaryTabChange = this.handlePrimaryTabChange.bind(this)
    this.renderEMRTable = this.renderEMRTable.bind(this)
    GridCell.role = this.props.role
    GridCell.fullName= this.props.fullName
    GridCell.postAddSteps = this.postAddSteps.bind(this)
    GridCell.postTerminateCluster = this.postTerminateCluster.bind(this)
    GridCell.postRotateAMI = this.postRotateAMI.bind(this)
    GridCell.postDNSFlip = this.postDNSFlip.bind(this);
    this.postCreateCluster= this.props.postCreateCluster
    this.handleActionBtnClick = this.handleActionBtnClick.bind(this);
    this.fetchAllClusterMetaData = this.fetchAllClusterMetaData.bind(this);
    this.fetchClusterWiseMetaData = this.fetchClusterWiseMetaData.bind(this);
    this.handleTabs = this.handleTabs.bind(this);
    this.getUIDropdownData = debounce(this.getUIDropdownData.bind(this), 2500,Â false);
    this.getRefreshButton = this.getRefreshButton.bind(this);
  }

  render() {
    const dropdownContent = (
      this.props.uiListData.accounts && <Menu>
        {
          this.props.uiListData.accounts.map(n => {
            return <MenuItem text={n} key={n} onClick={() => {
              this.account_range = n;
              this.props.handleRangeChange('accounts',n)
            }
          }/>
          })
        }
      </Menu>
    );
    this.handleTabs();
    return (
    <div className='emr-management-component'>
        {this.props.activeSpinner ? <div active={this.props} className='loader-wrapper'>
          <div className='loader' />
        </div>
        : null}
        <h2 style={{ marginLeft: '35px' }}>EMR <i className='material-icon title-arrow'>double_arrow</i> <span style={{ color: '#0097E4' }}>Orchestration </span><i className='material-icon title-arrow'>double_arrow</i><span style={{ color: '#7a7a7a' }}> {this.state.tabHeading}</span></h2>
        {/* Primary Tabs */}
          <div className='actions'>
            <ul>
              <li><button className={`actionBtn ${this.state.primaryAction == 'allClusters' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('allClusters')}><span>View Workflows</span></button></li>
              {this.showCreateTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'createCluster' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('createCluster')}><span>Create Cluster</span></button></li> : null}
              {this.showCreateTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'cloneCluster' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('cloneCluster')}><span>Clone Cluster</span></button></li> : null}
              {this.showTerminateTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'terminateCluster' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('terminateCluster')}><span>Terminate Cluster</span></button></li> : null}
              {this.showAddTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'addStep' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('addStep')}><span>Add Step</span></button></li> : null }
              {this.showRotateTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'rotateAMI' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('rotateAMI')}><span>Rotate AMI</span></button></li> : null}
              {this.showTestSuitesTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'testSuites' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('testSuites')}><span>Test Suites</span></button></li> : null }
              {this.showRotateTab ? <li><button className={`actionBtn ${this.state.primaryAction == 'dnsFlip' ? 'active1' : ''}`} onClick={() => this.handleActionBtnClick('dnsFlip')}><span>Flip To Production</span></button></li> : null }
            </ul>
          </div>
          <div className={`actionContent ${this.state.primaryAction == 'createCluster' ? 'actionContentCluster' : ''}`}>
            {this.state.primaryAction == 'createCluster' && !this.props.uiListFetching && this.props.uiListSuccess ?
              <div>
                {!this.props.globalJiraFetching && <CreateCluster 
                  {...this.props} 
                  clusterId={this.state.clusterId}
                /> }

              </div> :
              <div className='tabContent'>
                <Tabs className='line' id="primary-tab" selectedTabId={this.state.primaryTab} onChange={this.handlePrimaryTabChange}>
                  <Tab id="all" title="All Clusters" />
                  <Tab id="exploratory" title="Exploratory Clusters" />
                  <Tab id="scheduled" title="Schedule Clusters"/>
                  <Tab id="transient" title="Transient Clusters"/>
                  <Tab id="accounts">
                    <Popover content={dropdownContent} position={Position.BOTTOM}>
                      <li className="bp3-tab" role="tab" aria-selected="true">Select an Account</li>
                    </Popover>
                  </Tab>
            </Tabs>
            {
              this.props.fetching || this.props.uiListFetching ?
              <div className='loader-wrapper'>
                <div className='loader' />
              </div>
                :
                <div>
                  <ExcelExport
                  data={this.props.data}
                  ref={(exporter) => { this._export = exporter; }}
                >
                  {
                    this.renderEMRTable(this.state.primaryTab, this.shortlistData(this.props.data))
                  }
                </ExcelExport>
              </div>

            }
              </div>
            }

          </div>
        </div>
    )
  }

  renderEMRTable = (primaryTab, tableData) => {
    return (
      <div>
        <div>
          {
            this.props.role !== 'readonly' ?
            <div className="export-container">
              <button title='Download data into an excelsheet' className='exportBtn' onClick={this.export}>
                <i className='material-icons'>cloud_download</i>
                <span>Export to Excel</span>
              </button>
              {
                this.getRefreshButton(primaryTab)
              }
            </div>:
            <div/>
          }
        </div>
        {
          this.showTable(tableData)
        }
      </div>
    )
  }

  getRefreshButton = primaryTab => {
    switch (primaryTab) {
      case 'all':
        return (
          <button title='Refresh to get the latest data' className='exportBtn refresh' onClick={this.fetchAllClusterMetaData}>
              <i className='material-icons'>refresh</i>
              <span>Refresh Table</span>
          </button>
        )
      case 'exploratory':
        return (
          <button title='Refresh to get the latest data' className='exportBtn refresh' onClick={() => this.fetchClusterWiseMetaData('exploratory')}>
              <i className='material-icons'>refresh</i>
              <span>Refresh Table</span>
          </button>
        )
      case 'scheduled':
        return (
          <button title='Refresh to get the latest data' className='exportBtn refresh' onClick={() => this.fetchClusterWiseMetaData('scheduled')}>
              <i className='material-icons'>refresh</i>
              <span>Refresh Table</span>
          </button>
        )
      case 'transient':
        return (
          <button title='Refresh to get the latest data' className='exportBtn refresh' onClick={() => this.fetchClusterWiseMetaData('transient')}>
              <i className='material-icons'>refresh</i>
              <span>Refresh Table</span>
          </button>
        )
      case 'accounts':
        return (
          <button title='Refresh to get the latest data' className='exportBtn refresh' onClick={() => this.props.handleRangeChange('accounts', this.account_range)}>
              <i className='material-icons'>refresh</i>
              <span>Refresh Table</span>
          </button>
        )
    }
  }

  showTable = tableData => {
    let EMR_ID = <span style={{ fontSize: '15px', fontWeight: '600' }}>EMR ID</span>
    let EMR_Name = <span style={{ fontSize: '15px', fontWeight: '600' }}>EMR Name</span>
    let Cluster_Status = <span style={{ fontSize: '15px', fontWeight: '600' }}>Cluster Status</span>
    let Account_ID = <span style={{ fontSize: '15px', fontWeight: '600' }}>Account ID</span>
    let Created_by = <span style={{ fontSize: '15px', fontWeight: '600' }}>Created By</span>
    let Creation_Timestamp = <span style={{ fontSize: '15px', fontWeight: '600' }}>Created At</span>
    let Actions = <span style={{ fontSize: '15px', fontWeight: '600' }}>Actions</span>
    let Workflows = <span style={{ fontSize: '15px', fontWeight: '600' }}>Workflows</span>
    let AMIRotationDays = <span style={{ fontSize: '15px', fontWeight: '600' }}>AMI Rotation Days</span>
    let AutoPilot = <span style={{ fontSize: '15px', fontWeight: '600' }}>Auto Pilot Window & Days</span>
    let DNS_Name = <span style={{ fontSize: '15px', fontWeight: '600' }}>DNS Name</span>
    let JIRA_ID = <span style={{ fontSize: '15px', fontWeight: '600' }}>JIRA ID</span>
     
    let allClusters_id_WIDTH = "200px"
    let rotateAMI_id_WIDTH = "145px"
    let default_id_WIDTH = "260px"

    let allClusters_name_WIDTH = "325px"
    let rotateAMI_name_WIDTH = "190px"
    let default_name_WIDTH = "440px"


    let allClusters_status_WIDTH = "145px"
    let rotateAMI_status_WIDTH = "224px"
    let default_status_WIDTH = "150px"
    
    let rotateAMI_account_WIDTH = "120px"
    let default_account_WIDTH = "160px"

    let default_createdBy_WIDTH = "140px"
    let rotateAMI_createdBy_WIDTH = "120px"

    let default_createdAt_WIDTH = "115px"
    let rotateAMI_createdAt_WIDTH = "115px"
    let default_RotationDays_WIDTH = '130px'
    return (
      <StatefulGrid
              data={tableData}
              detail={(props) => <StepsStatus {...props} stepsStatusData={this.props.stepsStatusData} />}
              resizable
              groupable={true}
              onDataStateChange={this.dataStateChange}
              {...this.state.dataState}
              onExpandChange={this.expandChange}
              expandField="expanded"
              style={{ height: '80vh', fontSize: '13px' }}
          >
          {this.state.primaryAction === 'terminateCluster' ? <Column field="clusterId" title={EMR_ID} width={allClusters_id_WIDTH} /> : 
            this.state.primaryAction === 'rotateAMI' || this.state.primaryAction === 'allClusters' ? <Column field="clusterId" title={EMR_ID} width={rotateAMI_id_WIDTH} /> : 
             <Column field="clusterId" title={EMR_ID} width={default_id_WIDTH} />
          }
          {
            this.state.primaryAction === 'allClusters' ? <Column field="clusterName" title={EMR_Name} width='170px' /> :
          this.state.primaryAction === 'terminateCluster' ?  <Column field="clusterName" title={EMR_Name} width={allClusters_name_WIDTH} />  : 
            this.state.primaryAction === 'rotateAMI' ?  <Column field="clusterName" title={EMR_Name} width={rotateAMI_name_WIDTH} />  : 
            this.state.primaryAction === 'dnsFlip' ? <Column field="clusterName" title={EMR_Name} width="240px" />  :
            <Column field="clusterName" title={EMR_Name} width={default_name_WIDTH} /> 
          }

          {
            this.state.primaryAction === "rotateAMI" || this.state.primaryAction === 'allClusters' ? <Column field="status" title={Cluster_Status} width={allClusters_status_WIDTH} cell={emrStatusCell} /> : 
            <Column field="status" title={Cluster_Status} width={default_status_WIDTH} cell={emrStatusCell} />
          }     

          {this.state.primaryAction === 'rotateAMI' || this.state.primaryAction === 'allClusters' ?  <Column field="account" title={Account_ID} width={rotateAMI_account_WIDTH}/> : <Column field="account" title={Account_ID} width={default_account_WIDTH}/>}

          {(this.state.primaryAction === 'allClusters' || this.state.primaryAction === 'dnsFlip') && <Column field="dnsName" title={DNS_Name} width='200px' /> }
          
          {this.state.primaryAction === 'allClusters' && <Column field="requestTicket" title={JIRA_ID} width='110px' /> }

          {this.state.primaryAction === 'rotateAMI' ? <Column field="createdBy" title={Created_by} width={rotateAMI_createdBy_WIDTH} /> :
          <Column field="createdBy" title={Created_by} width={default_createdBy_WIDTH} /> }

          {this.state.primaryAction === 'rotateAMI' ? <Column field="creationTimestamp" title={Creation_Timestamp} width={rotateAMI_createdAt_WIDTH} /> :
          <Column field="creationTimestamp" title={Creation_Timestamp} width={default_createdAt_WIDTH} /> }

          {this.state.primaryAction === "rotateAMI" ? <Column field='rotationDaysToGo' title={AMIRotationDays} width={default_RotationDays_WIDTH} cell={amirotationStatus}/> : null}

          {this.state.primaryAction === 'rotateAMI' ? <Column field="actions" title={AutoPilot} width="230px" cell={(props) => <SubmitActionButtonCell {...props} {...this.props} action='autopilot' tab={this.state.primaryTab} handleActionBtnClick={this.handleActionBtnClick} />}/> : null }

          {this.state.primaryAction === "allClusters" ? <Column field="actions" title={Workflows} width="auto" cell={(props) => <SubmitActionButtonCell {...props} {...this.props} action={this.state.primaryAction} tab={this.state.primaryTab} handleActionBtnClick={this.handleActionBtnClick} />}/> :
          <Column field="actions" title={Actions} width="auto" cell={(props) => <SubmitActionButtonCell {...props} {...this.props} action={this.state.primaryAction} tab={this.state.primaryTab} handleActionBtnClick={this.handleActionBtnClick} />}/> }
        </StatefulGrid>
    )
  }

  handleTabs = () => {
    if (this.props.superAdmin) {
      this.showCreateTab = true;
      this.showAddTab = true;
      this.showTerminateTab = true;
      this.showRotateTab = true;
      this.showTestSuitesTab = true;
      return;
    }
    let emrRoles = this.props.dccRoles.filter(x => x.serviceType === 'EMR')[0];
    if (emrRoles !== undefined) {
      emrRoles.roles.forEach( x => {
        switch(x.name) {
          case 'createcluster':
            this.showCreateTab = true;
            break;
          case 'addstep':
            this.showAddTab = true;
            break;
          case 'terminatecluster':
            this.showTerminateTab = true;
            break;
          case 'rotateami':
            this.showRotateTab = true;
            break;
          case 'runclusterhealthchecks':
            this.showTestSuitesTab = true;
            break;
        }
      })
    }
  }

  shortlistData = data => {
    if (this.props.superAdmin) return data;
    let emrRoles = this.props.dccRoles.filter(x => x.serviceType === 'EMR')[0];

    if (emrRoles === undefined) return [];

    let roleSegments = [];
    emrRoles.roles.forEach((value) => {
      if (this.state.primaryAction === 'terminateCluster') {
        if (value.name === 'terminatecluster') {
          roleSegments = value.segments;
        }
      } else if (this.state.primaryAction == 'rotateAMI' || this.state.primaryAction == 'dnsFlip') {
        if (value.name == 'rotateami') {
          roleSegments = value.segments;
        }
      } else if (this.state.primaryAction === 'addStep') {
        if (value.name === 'addstep') {
           roleSegments = value.segments;
        }
      } else if (this.state.primaryAction === 'testSuites') {
        if (value.name === 'runclusterhealthchecks') {
          roleSegments = value.segments;
        }
      } else if (this.state.primaryAction === 'cloneCluster') {
        if (value.name === 'createcluster') {
          roleSegments = value.segments;
        }
      } else {
        value.segments.forEach(x => {
          roleSegments.push(x);
        })
      }
    });

    let roleMap = {};
    roleSegments.forEach((value, index) => {
      let accounts = [];
      value.accounts.forEach((act, index) => {
        accounts.push(act.accountId);
      })
      roleMap[value["segmentName"].toLowerCase()] = accounts;
    })
    let newData = data;
    
    newData = newData.filter((value) => {
      if (value.segment in roleMap) {
        if (roleMap[value.segment].includes(value.account)) {
          return value;
        }
      }
    });
    return newData;
  }

  fetchAllClusterMetaData() {
    console.log("Refresh all data");
    this.props.fetchAllClusterMetaData(this.props.token);
  }

  fetchClusterWiseMetaData(type) {
    console.log("Refreshing " + type + " data");
    this.props.fetchClusterWiseMetaData(type, this.props.token);
  }

  postAddSteps(data,token) {
    // this.props.postAddSteps(data,token)
    this.props.handleSpinnerActivity(data, token, 'add')
  }

  postTerminateCluster(data,token) {
    // this.props.postTerminateCluster(data,token)
    this.props.handleSpinnerActivity(data, token, 'terminate')
  }

  postRotateAMI(data,token) {
    // this.props.postRotateAMI(data,token)
    this.props.handleSpinnerActivity(data, token, 'rotate')
  }

  postCreateCluster(data,token) {
    // this.props.postCreateCluster(data,token)
    this.props.handleSpinnerActivity(data, token, 'create');
  }

  postDNSFlip(data, token) {
    this.props.handleSpinnerActivity(data, token, 'dnsFlip');
  }

  _export
  export() {
    this._export.save()
  }

  handlePrimaryTabChange(newTabId, prevTabId) {
    this.setState({
      ...this.state,
      primaryTab: newTabId
    })
    if (newTabId !== 'accounts')
      this.props.handleRangeChange(newTabId,'');

  }

  handleActionBtnClick(action, type, cluster_id) {
    let heading = '';
    if (action === 'allClusters') {
      heading = 'View All Clusters'
    } else if (action === 'createCluster') {
      heading = 'Create Cluster'
    } else if (action === 'addStep') {
      heading = 'Add Step'
    } else if (action === 'terminateCluster') {
      heading = 'Terminate Cluster'
    } else if (action === 'rotateAMI') {
      heading = 'Rotate AMI'
    } else if (action === 'testSuites') {
      heading = 'Run Test Suites'
    } else if (action === 'dnsFlip') {
      heading = 'DNS Flip'
    } else if (action === 'cloneCluster') {
      heading = 'Clone Cluster'
    }

    if (type == 'clone') {
      this.setState({
        ...this.state,
        primaryAction: action,
        tabHeading: heading,
        clusterId: cluster_id
      })
    } else {
      this.setState({
        ...this.state,
        primaryAction: action,
        tabHeading: heading
      })
    }
  }

  createGroupState(dataState) {
    //console.log('inside createGroup', this.props);
    return {
      result: process(this.props.data, dataState),
      dataState: dataState
    }
  }

  dataStateChange(event) {
    this.setState({
      ...this.state,
      ...this.createGroupState(event.data)
    })
  }

  expandChange(event) {
    event.dataItem.expanded = event.value;
    let cluster = event.dataItem.clusterName
    let clusterId = event.dataItem.clusterId
    this.setState({
      ...this.state,
      result: Object.assign({}, this.state.result),
      dataState: this.state.dataState
    });
    event.value && this.props.handleStepsStatus(cluster,clusterId);

    if (!event.value || event.dataItem.stepsStatusData) {
      return;
    }
  }
  getUIDropdownData() {
  }

  componentDidUpdate(prevProps) {
    if (prevProps.data && this.props.data && prevProps.data !== this.props.data) {
      this.setState({
        ...this.state,
        ...this.createGroupState({
          take: 100,
          group: Array(0),
          sort: undefined,
          filter: undefined,
          skip: 0
        })
      })
    }

    clearInterval(this.interval)

    if (this.state.primaryTab === "all") {
      this.interval = setInterval(this.fetchAllClusterMetaData, CLUSTER_REFRESH_TIME)
    } else if (this.state.primaryTab === "exploratory") {
      this.interval = setInterval(() => this.fetchClusterWiseMetaData('exploratory'), CLUSTER_REFRESH_TIME)
    } else if (this.state.primaryTab === "scheduled") {
      this.interval = setInterval(() => this.fetchClusterWiseMetaData('scheduled'), CLUSTER_REFRESH_TIME)
    } else if (this.state.primaryTab === "transient") {
      this.interval = setInterval(() => this.fetchClusterWiseMetaData('transient'), CLUSTER_REFRESH_TIME)
    } else if (this.state.primaryTab === "accounts") {
      this.interval = setInterval(() => this.props.handleRangeChange('accounts', this.account_range), CLUSTER_REFRESH_TIME)
    }


  }

  componentDidMount(prevProps) {
    this.setState({
      ...this.state,
      ...this.createGroupState({
        take: 100,
        group: Array(0),
        sort: undefined,
        filter: undefined,
        skip: 0
      })
    })
    this.props.fetchUIDropdownList(this.props.token)
    console.log('Comp EMR Managemnent', this)
  }

  componentWillReceiveProps(props, state) {
    this.forceUpdate();
  }
}
