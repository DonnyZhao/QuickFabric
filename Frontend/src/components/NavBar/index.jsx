import React from 'react';
import { NavLink } from 'react-router-dom';
import qdlogo from '../../assets/navbar/qdlogo.png';
import './navBar.scss';
import dcc_logo from '../../assets/navbar/qf_white.svg';

/**
 * Side navigation bar component.
 */
export default class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            dmvDisabled: true,
            emrDisabled: true,
        }
        this.handleDmvClick = this.handleDmvClick.bind(this);
        this.handleEmrClick = this.handleEmrClick.bind(this);
    }

    render() {
        return (
            <div className="sidebar"> 
                <NavLink to='/'><div className="brand">
                    <img src={qdlogo} alt="logo" className="logo"/>
                    <img src={dcc_logo} alt="logo" className="dcclogo"/>
                </div></NavLink>
                <ul className="appMenu">
                    <li className='menu-nav'><i className="material-icons">attach_money</i><span>EMR</span><i className="material-icons">keyboard_arrow_down</i></li>
                    <li><NavLink className='menu-nav submenu' to='/emrHealth'><span>Observability</span></NavLink></li>
                    <li><NavLink className='menu-nav submenu' to='/emrManagement'><span>Orchestration</span></NavLink></li>
                    <li><NavLink className='menu-nav submenu' to='/emrCost'><span>Cost</span></NavLink></li>
                    {this.props.superAdmin ? <li><NavLink className='menu-nav' to='/admin'><i className="material-icons">people</i><span>Administration</span></NavLink></li> : null}
                    
                </ul>
                <ul className="userMenu">
                    <li><NavLink className='menu-nav' to='/profile'><i className='material-icons'>account_circle</i><span>{this.props.fullName}</span></NavLink></li>
                    <li><NavLink className='menu-nav' to='/help'><i className="material-icons">help_outline</i><span>Help</span></NavLink></li>
                </ul>
            </div>
        )
    }

    handleDmvClick() {
        let val = this.state.dmvDisabled;
        this.setState({
            ...this.state,
            dmvDisabled: !val
        })
    }

    handleEmrClick() {
        let val = this.state.emrDisabled;
        this.setState({
            ...this.state,
            emrDisabled: !val
        });
    }
}