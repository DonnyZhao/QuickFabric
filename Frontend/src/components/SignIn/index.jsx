import React, { Component } from 'react'
import { FormGroup } from '@blueprintjs/core'
import './SignIn.css'
import baseURL from '../../api-config'
import DCCLogo from '../../assets/login/qf_black.svg';
import GithubLogo from '../../assets/login/gitLogo.png';

/**
 * Authenticate and sign in the user.
 */
class SignIn extends Component {
  constructor(props) {
    super(props)

    this.state = {
      username: '',
      passcode: '',
      displayHelperUsername: false,
      displayHelperPasscode: false
    }

    this.handleUsername = this.handleUsername.bind(this)
    this.handlePasscode = this.handlePasscode.bind(this)
    this.handleContinue = this.handleContinue.bind(this)
    this.handleKeyDown = this.handleKeyDown.bind(this)
  }

  render() {
    return (
      <div className="loginBg">
        <img className='dcc-logo' src={DCCLogo} />
        <div className='aws-video'>
          <iframe width="840" height="472.5" src="https://www.youtube.com/embed/TfgNzzaakNs?start=1324" frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowFullScreen></iframe>
        </div>
        <div className='sign-in-cont'>
          <div className='sign-in-component'>
            <div className="sign-in-card" onKeyDown={this.handleKeyDown} tabIndex="0">
              <a href={baseURL+'/login/sso/redirect'}> 
                <button disabled className="sign-in-card-button" onClick={this.handleContinue}> SSO Login </button>
              </a>
              <div className="login__saml-separator">
              <span>
                or
              </span>
              </div>
              <FormGroup 
                labelFor="username" 
                helperText={this.state.displayHelperUsername ? 'Username is required' : ''}
                className="inputForm"
              >
                <input className="inputField" value={this.state.username} placeholder='Username' onChange={this.handleUsername} id="username"/>
              </FormGroup>
              <FormGroup 
                labelFor="passcode" 
                helperText={this.state.displayHelperPasscode ? 'Passcode is required' : ''}
                className="inputForm"
              >
                <input className="inputField" value={this.state.passcode} placeholder='Password' type="password" onChange={this.handlePasscode} id="passcode"/>
              </FormGroup>
              <button className="dev-login-button" onClick={this.handleContinue}>Dev Login</button>
              {/* </a> */}
              <a href='https://github.com/intuit/QuickFabric' target='_blank'><img className='git-logo' src={GithubLogo} /></a>
            </div>
          </div>
        </div>
      </div>
    )
  }
  handleKeyDown(e) {
    if(e.keyCode == 13 && e.shiftKey == false) {
      e.preventDefault()
      this.handleContinue()
    }
  }
  handleUsername(e) {
    this.setState({
      ...this.state,
      username: e.target.value,
      displayHelperUsername: !(e.target.value)
    })
  }

  handlePasscode(e) {
    this.setState({
      ...this.state,
      passcode: e.target.value,
      displayHelperPasscode: !(e.target.value)
    })
  }

  handleContinue() {
    let { username, passcode } = this.state
    if (username && passcode ) {
      this.props.handleSignIn(username, passcode, null)
      this.setState({ username: '', passcode: ''})
      return
    }
    let newState = {}
    if (username === '')
      newState.displayHelperUsername = true
    if (passcode === '')
      newState.displayHelperPasscode = true
    this.setState({ ...this.state, ...newState })
  }
}

export default SignIn