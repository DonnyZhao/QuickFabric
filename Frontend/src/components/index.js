export {default as SignIn} from './SignIn'
export {default as EMRCost} from './EMRCost'
export {default as EMRHealth} from './EMRHealth'
export {default as EMRManagement} from './EMRManagement'
export {default as NavBar} from './NavBar'
export {default as Profile} from './Profile'
export {default as Admin} from './Admin'
export {default as Help} from './Help'