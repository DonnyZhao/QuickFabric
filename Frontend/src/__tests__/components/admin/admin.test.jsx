import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from "redux-mock-store";

import { Admin } from '../../../components';

const initialState = {
    uiListArray: {
        roles: [],
        accounts: [],
        response: {
            segments: []
        }
    }
}

const mockStore = configureMockStore();
const store = mockStore(initialState);
const component = shallow( 
    <Provider store={store}>
    <Admin {...initialState} />  
    </Provider>)    

describe('Testing for Admin', () => {
    it('Should render the component', () => {
        expect(toJson(component)).toMatchSnapshot();
    })
})