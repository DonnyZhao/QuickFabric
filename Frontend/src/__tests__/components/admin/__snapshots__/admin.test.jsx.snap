// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing for Admin Should render the component 1`] = `
<Connect(Admin)
  uiListArray={
    Object {
      "accounts": Array [],
      "response": Object {
        "segments": Array [],
      },
      "roles": Array [],
    }
  }
/>
`;
